### DISCUSSION TITLE
How does sensor.goe_xxx_pvopt_averagepgrid  get calculated?

### DISCUSSION THREAD

**Discussion by karliwalti** [marq24/ha-goecharger-api2#3] [Q&A]
My understanding is that this is the value that is being used to drive the controller of the PV optimized charging. The target would be a sensor.goe_xxx_pvopt_averagepgrid between -200 and 0 

The reason I ask is that I observe a strong oscillatory behavior in control if I transmit PV, Battery, and Grid information. When I send only grid information, the system runs much more stable but depletes the battery before stopping.

**Comment by marq24** [comment]
Hi Walt,

the simple & honest answer is - I don't have a clue at all. All data the integration provide will be read 1:1 from the Wallbox [technically this is not 100% correct - some of the values will be divided by a constant - you can find the factor attribute in the entity descriptions in the source code.

So the integration provide the data from the wall box without any kind of additional functionality. When you made the observation, that sending only GRID information would lead to more/stable results, then it could be the case, that the 'sign' of the battery might be not correct? This might be a a bug in the documentation [what means "exporting" and what means "importing" (from where to what?)]...

But coming back to your initial question -> 'sensor.goe_xxx_pvopt_averagepgrid' will be calculated by go-eCharger Wallbox internally and this integration just provide the data to HA... -> I just can recommend to ask your question @ https://github.com/goecharger/go-eCharger-API-v2/discussions

**Reply by karliwalti** [reply]
Hi @marq24 

OK, I think I get it now. This integration is a parser for the API and it also sends the values from the power sensors. My mistake was to think that the emulation of the eController would need more than that.
Thank you!

**Comment by marq24** [comment]
I am/was also totally surprised, that go-eCharger offers the feature to provide PV data "manually" via the API (as alternative to the installation of the hardware solution)... Actually when I saw this in the documentation @ github that was the point where I wanted to create/provide this integration - and also 'yes' you are fully right - this is a simple wrapper of the existing API into HA entities...

**Comment by karliwalti** [comment]
I think the truth is a bit in between. Reading the full posts here https://github.com/goecharger/go-eCharger-API-v2/discussions/110 suggests that pGrid is used, and to some extend also pAkku, but not pPv. this means that for more complex settings a custom controller on the HA side might be more suitable (by modifying/hacking pGrid and pAkku when in ECO mode or directly with the commands amp (Ladestrom einstellen), frc (Ladevorgang starten/beenden) und psm (Phasenumschaltung) )

**Reply by marq24** [reply]
IMHO the discussion thread you linked have indeed contradictory information - for me _only_ the posts from @0xFEEDC0DE64 are the 'source of information'... If you installed multiple chargers it will be for sure a total different story and additional hardware (or the use of evcc might be more sufficient -> https://github.com/marq24/ha-evcc).

I haven't spotted the post where you have found the info, that _"pAkku" should be used - and pPv should not be provided_... But at the end of the day it's not important - since you are in control of the automation (https://github.com/marq24/ha-goecharger-api2?tab=readme-ov-file#example-automation) and have the option to include __only__ the sensors as data provider you believe in.

Of course I am highly interested, if the situation will change for you, when you do not provide `ppv` data (or if the situation will remain the same)... We can also switch to German language here - since it might make things a bit easier in communication.

**Reply by karliwalti** [reply]
https://github.com/goecharger/go-eCharger-API-v2/discussions/110#discussioncomment-5533186
I dont have a reliable source for pAkku, but for pPV (only for display)

**Reply by IngmarStein** [reply]
So I also started using this integration including the automation which sends pPv, pAkku, and pGrid to the charger.
I noticed that the charger was overshooting, consuming the PV production, the maximum battery discharge, and importing from the grid to reach 11 kW and I think it's because of the pAkku direction. Like https://github.com/goecharger/go-eCharger-API-v2/discussions/110#discussioncomment-9554519, I've been passing through the values from a SolarEdge inverter which is positive when the battery is charging and negative when it's discharging. That script is negating the value and it was mentioned in the thread that negative values indicate charging, positive values discharging.

However, it's totally sensible to set pAkku to 0 (or not pass it) if one wants to prioritizes the home battery over the car.

**Reply by karliwalti** [reply]
Hi @IngmarStein  I use now following solution that seems to work quite well for my system / car  (note I have a battery with max 2500W C ). I set pAkku to 0 and modulate pGrid using following sensor template:

   ```
 - name: "Spare Power for Charging Car"
      unit_of_measurement: "W"
      state: > 
        {% set grid_active_power = (states('sensor.power_meter_active_power') | default(0) | float(0) * -1) %}
        {% set battery_active_power = (states('sensor.battery_charge_discharge_power_with_loss') | default(0) | float(0) * -1) %}
        {% if battery_active_power < 0.0 %} 
            {# battery is charging, add margin #}
            {% if battery_active_power < -2000.0 %} 
                {# battery is charging over a threshold, expose excess #}
                {{ battery_active_power + 2000.0 + grid_active_power }}
            {% else %} 
                {{ grid_active_power + 0}}
            {% endif %}
        {% else %}
          {# battery is feeding, sum up power fed into house #}
          {{ battery_active_power + grid_active_power }}
        {% endif %}
```

**Reply by IngmarStein** [reply]
So I got official confirmation from go-e support that the charger is currently always prioritized over the battery, so your template above is the right way to do it the other way around. Eventually, they want to expose the prioritization as a setting in the API and app.

**Comment by IngmarStein** [comment]
Maybe I'm not seeing something. I now only submit pgrid an here's a case where I plug in the car at 21:00 after sunset when the house was drawing from the battery. Before plugging it in, pgrid was ~0, as expected:

![Bildschirmfoto 2024-06-08 um 21 08 03](https://github.com/marq24/ha-goecharger-api2/assets/490610/42d71be3-fa6f-49ef-9fcd-43600ec6bcd4)

Still, charging started with 4.2 kW (6A, 3 phases) even though pgrid was far away from the configured start (1400 W):

![Bildschirmfoto 2024-06-08 um 21 08 19](https://github.com/marq24/ha-goecharger-api2/assets/490610/efc16476-5ba9-48e0-b643-3485c5c29901)
![IMG_4581 (1)](https://github.com/marq24/ha-goecharger-api2/assets/490610/321fcfbb-f146-4d78-9756-2bbbfecc4d48)

I'll now try @karliwalti's approach of adjusting pgrid by the battery power, but as I understood the semantics of pgrid, this shouldn't be necessary, no?

**Reply by marq24** [reply]
>![IMG_4581 (1)](https://private-user-images.githubusercontent.com/490610/337927146-321fcfbb-f146-4d78-9756-2bbbfecc4d48.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc5MTU4MjAsIm5iZiI6MTcxNzkxNTUyMCwicGF0aCI6Ii80OTA2MTAvMzM3OTI3MTQ2LTMyMWZjZmJiLWYxNDYtNGQ3OC05NzU2LTJiYmJmZWNjNGQ0OC5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwOSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDlUMDY0NTIwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Yzc4ZTI5ZTljOWNhYjk0NDExZDBjYTk5MjI5MDQ4ZGZmNjdiZTZkMmE4NGExZGNhOTQ3Y2ZlNjM4YTdkZTBiMyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.z8LwTw_Hq4ixRglPqRtXyPdR8-l2hUD5n7dGs-dME_k)
> 

With a home battery you should set the value `Zielwert Netz` to -200W (or lower) in your go-eCharger settings (that's a hint in the go-eCharger app)...

**Reply by IngmarStein** [reply]
Yeah, that's where I have it now. But even in the app, the description of the positive/negative direction of this value is confusing. It says "positive values lead to a lower grid import, negative values to a lower grid export". I would expect it to be the other way around.

**Reply by marq24** [reply]
I fully agree with you that the "instructions" are very confusing - and I also agree, that this "minus" (and minus minus minus means plus) is making me scratching my head all the time.

I just can assume, that the developers of the go-eCharger firmware have a technical point of view of these values - and for me it would help extremely to see the actual code that would do the math/calculation. Since in my personal experience sometimes alternative variable names and inverting values and the meaning would make the readability much easier... 

So let's see, what will happen with -200W once the sun goes down this evening (and the home battery is full) and you will plug in the car (and if it's not fully charged)... that's some "if's",,,

**Comment by karliwalti** [comment]
This is now a bit speculative, but it is likely that the charger "probes" how much it can drain until pGrid becomes positive (or larger Zielwert Netz). When a battery is attached, it can do this up to the max battery discharge current without seeing a change in the pGrid value. The default setting is likely that a battery discharge for charging a car is OK (since it is temporary stored PV power).

**Reply by IngmarStein** [reply]
I think that's plausible once charging has started.
When I sent the battery values according to the documentation, the Gemini interpreted that as "battery is charging" instead of "battery is draining" and then probably thought that there is a PV surplus that should go into the car instead of the house battery.
However, when I stopped sending pAkku, the go-e system didn't know about the battery and I can't explain why it started charging  before reaching the initial minimum of 1.4 kW (and why 3 phases).

Today, I tested how it looked with the fixed pAkku direction and during the day everything was as expected. That, however, prevented testing the scenario from yesterday, as the car battery is now full. Next time…

**Comment by core66** [comment]
I don't have a battery and had the same problem. I have now set this target grid value (Zielwert Netz) to -1500 and it now leaves me with 1500 watts surplus.

Is it working so far, am I right?

**Reply by marq24** [reply]
> am I right?

that's my understanding so far - can do some additional/internal tests end of next week...

**Comment by core66** [comment]
does anyone know whether it is possible to change the PV surplus calculation interval? it would be great if it could be reduced slightly rather than at 300 (stopped charging process) or 30 (active, paused charging process) second intervals.

**Comment by core66** [comment]
The main problem is actually the calculation interval. can this be changed anywhere?

**Reply by marq24** [reply]
this is a question you have to ask the the official go-eCharger Support team - since the integration can only adjust what's available via the API - and if you do not find a value in the API-Key documentation of go-e, then the chances are quite high, that currently it's not possible for the end user to adjust this interval...

**Comment by karliwalti** [comment]
I am not sure whether I can follow your question. Are you asking whether it is possible to override the pause?
Check here https://github.com/goecharger/go-eCharger-API-v2/blob/main/apikeys-en.md, the mc* can fix the pause length or stop the pause.
However, if your question is  to speed up the power tracker itself, then your best option is to move the power tracking control algo to HA and do all calculations yourself. You would be updating then at the update rate of any of the input parameters.

**Reply by core66** [reply]
ty mate, i mean the "pv surplus calculation" ... standard its 30 Seconds....this i wanna change to 10 secs for example

**Reply by core66** [reply]
Because I don't have a battery, I have the following problem:

Sun is shining - car is charging 

Cloud is coming or a larger consumer in the house is activated - charging is stopped (because the calculation takes too long for the charger to adapt to the available PV power).

**Comment by karliwalti** [comment]
If you check the API documentation link, you will find for 'ids':  `Needed all 5 seconds. Can be read back within 10 seconds after set. `
Are you providing the values at that interval? If not, you could force it by HA. The values are averaged by the go-e "pvopt_averagePGrid" over a certain period, i.e. 30 s (I am not sure about that number), and if in HA the numbers do not change, then they will not be sent. The consequence is that the average number lags.

Nevertheless, your logic above suggests the opposite: You need to implement a larger filter (stronger smoothing) to avoid stopping the car's charging (but you will pull power from the grid). You do not need to update the calculation frequency; you need to update the value you feed to 'pGrid'.  
The API also has a tool for you: 'sh'. The hysteresis will define how much change, and in combination with 'fmt', you could probably find a good solution for you with only API calls.

**Reply by core66** [reply]
oh ok that seems a bit too "high" for me.... how can i implement this? could you take me by the hand? otherwise i'll have to get myself a GO-E COntroller.... :*-| how can i force HA to provide faster'?

**Reply by marq24** [reply]
the HA Automation Script have a frequency - when you copied it, then this is 5sec by default... 

But my understanding of your issue is, that you want that the go-eCharger react faster then in a 30sec window in the case, that the GRID value going to drop... So you want, that when the PV power drops (cause of a cloud), that then in the next 5sec the charger will adjust the loading process (based on the new (lower) available power)... 

IMHO a eController would not solve this problem (that it can up to 30sec till the charger will react based on the available PV)... I assume that the whole control loop (Regelkreis) could not be implemented faster (since there will be an additional delay depending on your connected car)... So IMHO you have to live with the fact, that when the PV output is changing, that the charger needs some time to adjust the loading process (and will consume some grid-power)...

**Reply by karliwalti** [reply]
I can only help further when I fully understand your goal. Is it:

- [ ] maximize the PV self-consumption,
- [ ] minimize the grid consumption, or
- [ ] keep continuously charging the car, but with minimal grid consumption?

Also, which power meter do you have, and how frequently do you obtain that value in HA?

**Reply by core66** [reply]
i have a Huawei Inverter and a DTSU666H.

I have read that the e-controller measures every 0.5 seconds and adjusts the level.


**Here is the excerpt from the post:** 

I wanted to share my experiences with the go-e controller with you here. This is basically the hardware part for overcharging. I had it installed yesterday by a friend. It was quite easy, took a maximum of 1 hour, and yesterday I immediately activated the surplus charging. This morning, as soon as there was power, the charger started as required and ramped up the power accordingly. The phase switching also worked. When the first car was at full 11 kw charging power, I activated the second. Also works perfectly. The power is distributed to both. It is regulated in approx. 0.5 seconds - much faster than I could with HA. I don't think you can complain for 250 EUR.

Here is another screenshot:

![1d0edcbf2737045bc947a01d846a9177c6c48f1f](https://github.com/marq24/ha-goecharger-api2/assets/37327446/f538dabd-b7dc-466b-9df7-ab7bd2af9f1b)

**Reply by core66** [reply]
If you have a battery system, the integration works wonderfully because you never actually get into the "positive load range" cause the battery compensates the "overload gap".

**Reply by marq24** [reply]
> I have read that the e-controller measures every 0.5 seconds and adjusts the level.

IMHO (just **my** opinion) - no matter how fast the eControler is measuring (and provide data), the only unit that can **adjust the level** is the eCharger... So if your 'Huawei Inverter and a DTSU666H' provide data every second, you can send the current data (via the automation) to the charger... 

you can increase the update interval (for sending) by adjusting the automation trigger:

from
```
trigger:
  - platform: time_pattern
    seconds: /5
```
to
```
trigger:
  - platform: time_pattern
    seconds: /1
```
But this only makes sense, if your HA Integration for the PV will deliver fresh data in the same frequency! - the Question from @karliwalti
 is still valid - what is the goal you would like to archive?

**Reply by core66** [reply]
My goal is....that the charging process is not stopped if the load on the grid increases briefly and the value falls into the "positive" value before the charger (+ integration) has adapted to the new value.

**Reply by karliwalti** [reply]
Alright, some more questions then:
1. do have a e-controller or not? The e-controller would replace the DTSU666H and HA.
2. How do you integrate your Huawei system in HA? Are you using the serial or internal WLAN? https://github.com/wlcrs/huawei_solar  or are you using the FusionSolar cloud data i.e. https://github.com/tijsverkoyen/HomeAssistant-FusionSolar ?
In the earlier option, you have to set the polling frequency to 5s as shown in https://github.com/wlcrs/huawei_solar?tab=readme-ov-file#inverter-polling-frequency, in second option you robablyhttps://github.com/wlcrs/huawei_solar?tab=readme-ov-file#inverter-polling-frequency have to reconfigure the inverter to send more often to the cloud)
3. What is the max acceptable load on the grid and for how long?

**Reply by core66** [reply]
No i have no controller yet. i use the huawei_solar integration, polling is here 10 sec....but I'm thinking more in a different direction: isn't it somehow possible to set a kind of gap in the "stop automation" of Marq24 so that it doesn't stop immediately but only after one minute, for example?? because the adjustment happens after 30 seconds at the latest?

**Reply by marq24** [reply]
Well - if the _optional_ `go-e FORCE STOP of PV surplus charging` automation is the root cause - then simply disable it - there had been reports, that the charger did not stop charging when PV is gone (sun is down) - so this additional automation is really just optional.

**Reply by karliwalti** [reply]
I do not use this automation, it just works fine for me without...
 
If you want to use it or if the averaging in the pv surplus calculation of the charger is not sufficient, use a filter: 
https://www.home-assistant.io/integrations/filter/
the low pass or moving average will do what you want.

**Comment by core66** [comment]
ok the solution with the deactivation of the stop automation was good. Now I still have the problem that the recalculation of the surplus (the 30 seconds) is simply too long.

Example: If my car is charging with 22kW and then suddenly a cloud appears, it can happen that it continues to charge with 22kW for 25 seconds even though the value from the PV has long been less. I ask again: Is it not possible to change this calculation time???? I get the data from my inverter every 10 seconds.

**Reply by marq24** [reply]
> Is it not possible to change this calculation time???? 

only go-e can give you this answer - in the API there is no key to adjust this value - so this integration can't do this.

**Reply by karliwalti** [reply]
This is not an issue with this plugin but likely due to the rate (and filter) used by the charger. It could very well be (for good reason) that it only reacts slowly to changes. It is likely not very good for the switching electronics (in charger and car) to change more frequently than 30 seconds.  

If you want to ignore this warning, then here is your answer:
HA and the plugin recalculate and send new values to the charger as often as one value in the equation gets updated. In your case this will be 10s. In mine it is rather 5 s:
![image](https://github.com/marq24/ha-goecharger-api2/assets/4565568/9084a132-078d-4300-8185-4a1493b19fcb)

To overwrite the charger logic, you have to turn it on/off and set the load through the API.  That means all logic is run on HA.
1. Turn off PV charging on charger. 
2.  send start/stop, phase number and load

This will be a bit complicated, as you need to also decide when to do phase switch etc.
